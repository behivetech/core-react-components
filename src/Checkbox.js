// Vendor Libs
import React, {Component, createRef} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {noop} from 'lodash';
import {MDCCheckbox} from '@material/checkbox';
import {MDCFormField} from '@material/form-field';

// Components
import FormFieldHelperText from './FormFieldHelperText';

// Styles
import './Checkbox.scss';

export default class Checkbox extends Component {
    constructor(props) {
        super(props);

        if (!this.props.renderComponent) {
            this.checkboxRef = createRef();
            this.formFieldRef = createRef();
        }

        this.state = {checked: props.checked || false};
    }

    componentDidMount() {
        if (!this.props.renderComponent) {
            const checkbox = new MDCCheckbox(this.checkboxRef.current);
            const formField = new MDCFormField(this.formFieldRef.current);

            formField.input = checkbox;
        }
    }

    componentWillReceiveProps({checked}) {
        if (this.props.checked !== checked) {
            this.setState({checked});
        }
    }

    handleChange = (event) => {
        // It's unadvised to use handleChange because IE doesn't aways kick this off.
        // It's best to use onClick.
        event.stopPropagation();

        const {onChange} = this.props;

        if (onChange) {
            onChange(event);
        }
    }

    handleClick = (event) => {
        event.stopPropagation();
        this.setState({checked: event.target.checked || false});
        this.props.onClick(event);
    }

    getCheckboxClass() {
        return classnames({
            'mdc-checkbox': true,
            'mdc-checkbox--disabled': this.props.disabled,
        });
    }

    getInputProps = () => {
        const {disabled, name, style} = this.props;

        return {
            checked: this.state.checked,
            disabled,
            id: name,
            name: name,
            onChange: this.handleChange,
            onClick: this.handleClick,
            readOnly: true,
            style,
            type: 'checkbox',
            value: this.props.defaultValue,
        };
    }

    render() {
        const {
            className,
            error,
            helperText,
            label,
            name,
            renderComponent,
        } = this.props;

        return (renderComponent)
            ? renderComponent(
                this.handleClick,
                {...this.props, ...this.getInputProps()},
                this.state
            )
            : (
                <div className={classnames('checkbox', className)}>
                    <div className="mdc-form-field" ref={this.formFieldRef}>
                        <div className={this.getCheckboxClass()} ref={this.checkboxRef}>
                            <input
                                {...this.getInputProps()}
                                className="mdc-checkbox--input mdc-checkbox__native-control"
                            />
                            <div className="mdc-checkbox__background">
                                <svg className="mdc-checkbox__checkmark" viewBox="0 0 24 24">
                                    <path
                                        className="mdc-checkbox__checkmark-path"
                                        fill="none"
                                        d="M1.73,12.91 8.1,19.28 22.79,4.59"
                                    />
                                    </svg>
                                <div className="mdc-checkbox__mixedmark" />
                            </div>
                        </div>
                        <label className="checkbox--label mdc-checkbox--label" htmlFor={name}>{label}</label>
                    </div>
                    <FormFieldHelperText
                        className="checkbox--helper"
                        error={error}
                        helperText={helperText}
                        style={{marginLeft: '10px'}}
                    />
                </div>
            );
    }
}

Checkbox.propTypes = {
    className: PropTypes.string,
    checked: PropTypes.bool,
    /** The actual value of the checkbox. If not set, defaults to true */
    defaultValue: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
    ]).isRequired,
    disabled: PropTypes.bool,
    label: PropTypes.string,
    name: PropTypes.string,
    onClick: PropTypes.func,
    onChange: PropTypes.func,
    /**
        Function to give the ability to render different content for the checkbox.
        hanldeClick, props and state are passed in as arguments of the function.
        Primary example can be seen with MdcSwitch.
    */
    renderComponent: PropTypes.func,
    style: PropTypes.object,
    value: PropTypes.oneOfType(
        [PropTypes.arrayOf(PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.number,
            PropTypes.string,
        ])),
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
    ]),
};

Checkbox.defaultProps = {
    onClick: noop,
    onChange: noop,
};
