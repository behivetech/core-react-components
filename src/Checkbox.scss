@import "@material/form-field/mdc-form-field";
@import "@material/checkbox/mdc-checkbox";

.checkbox {
    display: inline-block;
    position: relative;
}
