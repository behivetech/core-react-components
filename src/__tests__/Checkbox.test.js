import Checkbox from "../Checkbox";

// TODO: need to figure out how to test MDCSwitch and MDCFormField
describe('Checkbox', () => {
    const getEvent = (checked = false) => { 
        return {
            stopPropagation: jest.fn(),
            target: {
                checked,
                value: 'mock event value',
            },
        }
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with the correct props', () => {
        const wrapper = shallow(<Checkbox name="mockName" className="mock-class" defaultValue="mock value" />);

        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should render correctly with the prop checked', () => {
        const wrapper = shallow(<Checkbox name="mockName" checked defaultValue="mock value" />);

        expect(wrapper.debug()).toMatchSnapshot();
        expect(wrapper.state().checked).toBe(true);
        wrapper.setProps({checked: false});
        expect(wrapper.state().checked).toBe(false);
    });

    it('should render as disabled', () => {
        const wrapper = shallow(<Checkbox name="mockName" checked defaultValue="mock value" />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should handle clicks correctly', () => {
        const wrapper = shallow(
            <Checkbox name="mockName" checked defaultValue="mock value" onClick={undefined} />
        );
        const wrapperInstance = wrapper.instance();
        const mockEvent = getEvent(true);
        const mockOnClick = jest.fn();

        expect(wrapperInstance.state.checked).toBe(true);
        wrapperInstance.handleClick(getEvent());
        expect(wrapper.state().checked).toBe(false);
        wrapper.setProps({onClick: mockOnClick})
        wrapperInstance.handleClick(mockEvent);
        expect(mockEvent.stopPropagation).toHaveBeenCalled();
        expect(mockOnClick).toHaveBeenCalledWith(mockEvent);
        expect(wrapperInstance.state.checked).toBe(true);
    });

    it('should handle changes correctly', () => {
        const mockOnChange = jest.fn();
        const mockEvent = getEvent();
        const wrapperInstance = shallow(
            <Checkbox name="mockName" checked defaultValue="mock value" onChange={mockOnChange} />
        ).instance();

        wrapperInstance.handleChange(mockEvent);
        expect(mockEvent.stopPropagation).toHaveBeenCalled();
        expect(mockOnChange).toHaveBeenCalledWith(mockEvent);
    });

    it('should not set refs for the component', () => {
        const wrapper = mount(
            <Checkbox name="mockName" defaultValue="mock value" renderComponent={() => null} />
        );
        const wrapperInstance = wrapper.instance();

        expect(wrapperInstance.checkboxRef).toBe(undefined)
        expect(wrapperInstance.formFieldRef).toBe(undefined)
    });

    it('should set refs for the component', () => {
        const wrapper = mount(<Checkbox name="mockName" defaultValue="mock value" />);
        const wrapperInstance = wrapper.instance();

        expect(wrapperInstance.checkboxRef).toMatchSnapshot();
        expect(wrapperInstance.formFieldRef).toMatchSnapshot();
    });
});