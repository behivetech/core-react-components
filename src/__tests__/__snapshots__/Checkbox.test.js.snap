// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render as disabled 1`] = `
<div
  className="checkbox"
>
  <div
    className="mdc-form-field"
  >
    <div
      className="mdc-checkbox mdc-checkbox--disabled"
    >
      <input
        className="mdc-checkbox__native-control"
        disabled={true}
        id="mockName"
        type="checkbox"
      />
      <div
        className="mdc-checkbox__background"
      >
        <svg
          className="mdc-checkbox__checkmark"
          viewBox="0 0 24 24"
        >
          <path
            className="mdc-checkbox__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          className="mdc-checkbox__mixedmark"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkbox should render correctly with the prop checked 1`] = `
<div
  className="checkbox"
>
  <div
    className="mdc-form-field"
  >
    <div
      className="mdc-checkbox"
    >
      <input
        className="mdc-checkbox__native-control"
        id="mockName"
        type="checkbox"
      />
      <div
        className="mdc-checkbox__background"
      >
        <svg
          className="mdc-checkbox__checkmark"
          viewBox="0 0 24 24"
        >
          <path
            className="mdc-checkbox__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          className="mdc-checkbox__mixedmark"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkbox should render with the correct props 1`] = `
<div
  className="checkbox mock-class"
>
  <div
    className="mdc-form-field"
  >
    <div
      className="mdc-checkbox"
    >
      <input
        className="mdc-checkbox__native-control"
        id="mockName"
        type="checkbox"
      />
      <div
        className="mdc-checkbox__background"
      >
        <svg
          className="mdc-checkbox__checkmark"
          viewBox="0 0 24 24"
        >
          <path
            className="mdc-checkbox__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          className="mdc-checkbox__mixedmark"
        />
      </div>
    </div>
  </div>
</div>
`;
